src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs

```diff
@@ -11,73 +11,150 @@ namespace Microsoft.CodeAnalysis.Differencing
-
-        protected IEnumerable<KeyValuePair<int, int>> GetMatchingPairs(TSequence oldSequence, int oldLength, TSequence newSequence, int newLength)
+        // VArray class enables array indexing in range [-d...d].
+        private class VArray
         {
-            int[,] d = ComputeCostMatrix(oldSequence, oldLength, newSequence, newLength);
-            int i = oldLength;
-            int j = newLength;
+            private int[] array;
+            private int offset;
 
-            while (i != 0 && j != 0)
+            public VArray(int d, VArray previousVArray) : this(d)
             {
-                if (d[i, j] == d[i - 1, j] + DeleteCost)
+                if (previousVArray != null)
                 {
-                    i--;
+                    int copyDelta= offset - previousVArray.offset;
+                    if (copyDelta >= 0)
+                    {
+                        Debug.Assert(previousVArray.array.Length + 2 * copyDelta == array.Length);
+                        Array.Copy(previousVArray.array, 0, array, copyDelta, previousVArray.array.Length);
+                    }
+                    else
+                    {
+                        Debug.Assert(previousVArray.array.Length + 2 * copyDelta == array.Length);
+                        Array.Copy(previousVArray.array, -copyDelta, array, 0, array.Length);
+                    }
                 }
-                else if (d[i, j] == d[i, j - 1] + InsertCost)
+            }
+
+            public VArray(int d)
+            {
+                offset = d;
+                array = new int[2 * d + 1];
+            }
+
+            public int this[int index]
+            {
+                get
                 {
-                    j--;
+                    return array[index + offset];
                 }
-                else
+                set
                 {
-                    i--;
-                    j--;
-                    yield return new KeyValuePair<int, int>(i, j);
+                    array[index + offset] = value;
                 }
             }
         }
 
-        protected IEnumerable<SequenceEdit> GetEdits(TSequence oldSequence, int oldLength, TSequence newSequence, int newLength)
+        protected abstract bool ItemsEqual(TSequence oldSequence, int oldIndex, TSequence newSequence, int newIndex);
+
+        protected IEnumerable<KeyValuePair<int, int>> GetMatchingPairs(TSequence oldSequence, int oldLength, TSequence newSequence, int newLength)
         {
-            int[,] d = ComputeCostMatrix(oldSequence, oldLength, newSequence, newLength);
-            int i = oldLength;
-            int j = newLength;
+            Stack<VArray> stackOfVs = ComputeEditPaths(oldSequence, oldLength, newSequence, newLength);
+
+            int x = oldLength;
+            int y = newLength;
 
-            while (i != 0 && j != 0)
+            for (int d = stackOfVs.Count - 1; x > 0 || y > 0; d--)
```

Comment:
I'd replace lines 66-68 with

```C#
while (x > 0 || y > 0)
{
   VArray currentV = stackOfVs.Pop();
   int d = stackOfVs.Count;

   ...
}
```

To avoid calculating stack depth in ```d``` as it's already counted by the stack itself.